File: ios/Sources/SduiCore/Rendering/ImageVariantResolver.swift
import SwiftUI
import os

private let logger = Logger(subsystem: "com.nba.sdui", category: "ImageVariantResolver")

/// Wire-level image variant vocabulary. Maps to a platform-native realization
/// via `ImageVariantSpec`. Unknown server strings fall through to the
/// primitive's default rendering.
enum ImageVariant: String {
    case thumbnail
}

/// Platform-native realization of an image variant for the iOS client.
/// Mirrors the iOS column of `schema/style-tokens.json`.
struct ImageVariantSpec {
    let cornerRadius: CGFloat?
    let aspectRatio: CGFloat?
    let contentMode: ContentMode
    let fillWidth: Bool
    let clip: Bool
    let overrideMatrix: [String: OverridePolicy]
}

enum ImageVariantResolver {

    /// Returns the realized spec for a recognized variant, or `nil` to let
    /// the primitive fall back to its default rendering.
    static func resolve(_ variant: String?, formFactor: String = RequestEnvelope.currentFormFactor) -> ImageVariantSpec? {
        guard let raw = variant, !raw.isEmpty else { return nil }
        guard let value = ImageVariant(rawValue: raw) else {
            logger.debug("unknown image variant \(raw, privacy: .public); rendering with primitive defaults")
            return nil
        }
        return spec(for: value, formFactor: formFactor)
    }

    private static func spec(for variant: ImageVariant, formFactor: String) -> ImageVariantSpec {
        let isTablet = formFactor == "tablet"
        switch variant {
        case .thumbnail:
            return ImageVariantSpec(
                cornerRadius: isTablet ? 12 : 8,
                aspectRatio: nil,
                contentMode: .fill,
                fillWidth: false,
                clip: true,
                overrideMatrix: [
                    "padding": .allow,
                    "cornerRadius": .allow,
                    "background": .allow,
                    "shadow": .allow,
                    "color": .allow,
                    "opacity": .allow,
                    "border": .allow
                ]
            )
        }
    }
}

File: android/sdui-core/src/main/java/com/nba/sdui/core/renderer/ImageVariantResolver.kt
package com.nba.sdui.core.renderer

import android.util.Log
import com.nba.sdui.core.renderer.ContainerVariantResolver.OverridePolicy

/**
 * Resolves a server-emitted image variant token (`thumbnail`) to a
 * platform-native spec the [AtomicImage] renderer applies to a Coil
 * AsyncImage. Each value in the vocabulary carries a platform-native
 * treatment inline props cannot cleanly express; values that reduce to
 * inline cornerRadius + aspectRatio + fit do not belong here.
 *
 * The registry values mirror `schema/style-tokens.json` — the JSON file is
 * the governance artifact; this object is the hand-authored snapshot the
 * Android client ships with.
 */
object ImageVariantResolver {

    private const val TAG = "ImageVariantResolver"

    /** Neutral content-scale vocabulary mapped to Compose `ContentScale` by caller. */
    enum class ImageContentScaleHint { Crop, Fit, FillBounds, None }

    data class ImageVariantSpec(
        val cornerRadiusDp: Int?,
        val aspectRatio: Float?,
        val contentScaleHint: ImageContentScaleHint,
        val fillWidth: Boolean?,
        /**
         * Whether to clip the image to its corner-radius shape. Logo imagery
         * must retain its native aspect ratio and not clip.
         */
        val clip: Boolean = true,
        val overrideMatrix: Map<String, OverridePolicy>
    )

    private val ALL_ALLOW: Map<String, OverridePolicy> = mapOf(
        "padding" to OverridePolicy.ALLOW,
        "cornerRadius" to OverridePolicy.ALLOW,
        "background" to OverridePolicy.ALLOW,
        "shadow" to OverridePolicy.ALLOW,
        "color" to OverridePolicy.ALLOW,
        "opacity" to OverridePolicy.ALLOW,
        "border" to OverridePolicy.ALLOW
    )

    /**
     * Resolve the variant token to an [ImageVariantSpec].
     *
     * @return `null` when [variant] is null / blank OR when it is a non-null
     *   value not present in the registry (a warning is logged and the
     *   renderer falls through to inline-only behavior).
     */
    fun resolve(variant: String?, formFactor: String = "phone"): ImageVariantSpec? {
        if (variant.isNullOrBlank()) return null
        return when (variant) {
            "thumbnail" -> {
                val isTablet = formFactor == "tablet"
                ImageVariantSpec(
                cornerRadiusDp = if (isTablet) 12 else 8,
                aspectRatio = null,
                contentScaleHint = ImageContentScaleHint.Crop,
                fillWidth = false,
                clip = true,
                overrideMatrix = ALL_ALLOW
            )
            }
            else -> {
                Log.w(TAG, "variant_resolver_missing: $variant")
                null
            }
        }
    }

    /** See [ContainerVariantResolver.logOverrideBlocked]. */
    fun logOverrideBlocked(variant: String?, axis: String, attemptedValue: Any?) {
        Log.d(
            TAG,
            "variant_override_blocked: variant=$variant axis=$axis attemptedValue=$attemptedValue"
        )
    }
}
